### PR TITLE
fix: when the vote expired, set federation status to failed.

### LINKED ISSUE #235: 提议管理-提议超时应返回失败信息
创建联盟时，初始成员未操作提议投票，导致提议超时，创建联盟状态一直是“组建中”，应该为“组建失败”

image
image

### PR DESCRIPTION
**Pull request by 0xff-dev**
Fix: https://github.com/bestchains/fabric-operator/issues/235

### CODE REVIEW

File: api/v1beta1/federation.go [hunk 1]
```diff
@@ -2,6 +2,12 @@ package v1beta1
 
 import "os"
 
+const (
+	KIND                               = "Federation"
+	FEDERATION_INITIATOR_LABEL         = "bestchains.federation.initiator"
+	FEDERATION_CREATION_PROPOSAL_ENDAT = "bestchains.federation.creation.proposal.endat"
```
Comment (bjwswang) [inline api/v1beta1/federation.go]: use `endAt` ?

File: api/v1beta1/proposal_webhook.go [hunk 2]
```diff
@@ -259,3 +264,21 @@ func validateDeleteFedMember(ctx context.Context, c client.Client, deleteMember,
 	}
 	return nil
 }
+
+func AddEndAtAnnotation(r client.Reader, p Proposal) (*Federation, error) {
+	fed := &Federation{}
```
Comment (bjwswang) [inline api/v1beta1/proposal_webhook.go:269]: we should check whether this label exists. If yes, we should block this proposal creation.

File: controllers/federation/predict.go [hunk 3]
```diff
@@ -157,6 +157,22 @@ func (r *ReconcileFederation) UpdateFunc(e event.UpdateEvent) bool {
...
 
+func (r *ReconcileFederation) ProposalCreateFunc(e event.CreateEvent) bool {
+
+	p := (e.Object).(*current.Proposal)
+	fed, err := current.AddEndAtAnnotation(r.client, *p)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to get federation %s", p.Spec.Federation))
+		return false
+	}
+	if err := r.client.Update(context.TODO(), fed); err != nil {
```
Comment (bjwswang) [inline controllers/federation/predict.go:168]: why add this annotation again?

File: api/v1beta1/proposal_webhook.go [hunk 4]
```diff
@@ -62,6 +62,11 @@ func (r *Proposal) Default(ctx context.Context, client client.Client, user authe
 	if r.Spec.EndAt.IsZero() {
 		r.Spec.EndAt = metav1.NewTime(time.Now().Add(time.Hour * 24))
 	}
+	if fed, err := AddEndAtAnnotation(client, *r); err == nil {
+		if e := client.Update(context.TODO(), fed); e != nil {
```
Comment (Abirdcfly) [inline api/v1beta1/proposal_webhook.go:66]: In Default(), there is no need to update?

Comment (0xff-dev) [reply api/v1beta1/proposal_webhook.go:66]: The webhook is for proposal, the update is  federation, so it should still be needed.

File: controllers/federation/predict.go (hunk 3 above)
Comment (0xff-dev) [reply controllers/federation/predict.go:168]: This is for cases where there is no webhook.

File: api/v1beta1/proposal_webhook.go (hunk 2 above)
Comment (0xff-dev) [reply api/v1beta1/proposal_webhook.go:269]: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks
According to the documentation, **the mutatingwebhook will be called first**, where the relevant annotation has been added to the federation, and the logic is incorrect when judged by the validatingwebhook.

If we want to do validation on the creation of a proposal for a federation type, then the logic for adding annotations should not be done at the wehbook level.

File: api/v1beta1/federation.go (hunk 1 above)
Comment (bjwswang) [reply api/v1beta1/federation.go]: @0xff-dev how about this annotation name?

**Review by bjwswang — APPROVED** [review]
/lgtm

File: api/v1beta1/proposal_webhook.go [hunk 5]
```diff
@@ -259,3 +264,21 @@ func validateDeleteFedMember(ctx context.Context, c client.Client, deleteMember,
...
+func AddEndAtAnnotation(r client.Reader, p Proposal) (*Federation, error) {
+	fed := &Federation{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: p.Spec.Federation}, fed); err != nil {
+		proposallog.Error(err, fmt.Sprintf("failed to get federation %s with porposal %s", p.Spec.Federation, p.GetName()))
+		return nil, err
+	}
+	if fed.Annotations == nil {
+		fed.Annotations = make(map[string]string)
+	}
+	if !p.Spec.EndAt.IsZero() {
```
Comment (Abirdcfly) [inline api/v1beta1/proposal_webhook.go]: maybe should check proposal type == createFederation?

**Review by Abirdcfly — APPROVED** [review]

**Pull request merged**